supervision/video/backends/opencv.py

```diff
@@ -0,0 +1,195 @@
+from collections.abc import Callable, Iterator
+from typing import Any
+
+import cv2
+import numpy as np
+
+from ..utils import VideoInfo
+from .base import Writer
+
+
+class OpenCVWriter:
+    def __init__(self, vw: cv2.VideoWriter, info: VideoInfo):
+        self._vw = vw
+        self.info = info
+
+    def write(
+        self,
+        frame: np.ndarray,
+        frame_number: int,
+        callback: Callable[[np.ndarray], None] | None = None,
+    ) -> None:
+        if callback:
+            frame = callback(frame, frame_number)
+        if frame.shape[0] != self.info.height or frame.shape[1] != self.info.width:
+            frame = cv2.resize(frame, (self.info.width, self.info.height))
+        self._vw.write(frame)
+
+    def close(self) -> None:
+        self._vw.release()
+
+    def __enter__(self) -> Writer:
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        self.close()
+
+
+class Backend:
+    def __init__(self, source_path: str | int):
+        """Create a new backend for source.
+
+        `source`` can b
+        * ``str`` - file path, RTSP/HTTP URL …
+        * ``int`` - webcam index (OpenCV-style)
+        """
+        self.source_path = source_path
+        self.cap = cv2.VideoCapture(self.source_path)
+        if not self.cap.isOpened():
+            raise ValueError(f"Could not open video source {self.source_path}")
+
+    def info(self) -> VideoInfo:
+        """Return static information (width / height / fps / total_frames)."""
+        from ..core import VideoInfo
+
+        w = int(self.cap.get(cv2.CAP_PROP_FRAME_WIDTH))
+        h = int(self.cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
+        precise_fps = self.cap.get(cv2.CAP_PROP_FPS)
+        fps = int(round(precise_fps, 0))
+        n = int(self.cap.get(cv2.CAP_PROP_FRAME_COUNT))
+        return VideoInfo(w, h, fps, precise_fps, n)
+
+    def read(self) -> tuple[bool, np.ndarray]:
+        """Decode the next frame."""
+        return self.cap.read()
+
+    def grab(self) -> bool:
+        """Grab the next frame without decoding pixels."""
+        return self.cap.grab()
+
+    def seek(self, frame_idx: int) -> None:
+        """Seek to frame_idx so that the next :py:meth:`read` returns it."""
+        self.cap.set(cv2.CAP_PROP_POS_FRAMES, frame_idx)
+
+    # ? Do we want to mix and match different writers to different backends?
```

Comment:
The writer should be consistent with the backend. For example, `OpenCVBackend` should use `OpenCVWriter`, and so on.